### PR TITLE
Fix issue with token-based locking

### PR DESCRIPTION
**Pull request by kostko**

### CODE REVIEW

File: channels_redis/core.py [hunk 1]
```diff
@@ -241,17 +241,38 @@ def _setup_encryption(self, symmetric_encryption_keys):
...
+                        break
+                    else:
+                        assert False
+
                     try:
                         message_channel, message = await self.receive_single(real_channel)
                         await self.receive_buffer[message_channel].put(message)
+                        message = None
                     finally:
                         await self.receive_tokens.put(token)
```
Comment (jberci) [inline channels_redis/core.py]: If the message got there first, that means the token waiter will be cancelled, so `token` won't have any meaningful value. It must not be returned back to the token queue in that case, as we could subsequently have two concurrent getters on the input channel.

Comment (kostko) [reply channels_redis/core.py]: But if we didn't get the token we break so this shouldn't run?

File: channels_redis/core.py [hunk 2]
```diff
@@ -231,36 +230,82 @@ def _setup_encryption(self, symmetric_encryption_keys):
...
+                            self.receive_lock.release()
 
                 # We know there's a message available, because there
                 # couldn't have been any interruption between empty() and here
-                message = self.receive_buffer[channel].get_nowait()
+                if message is None:
+                    message = self.receive_buffer[channel].get_nowait()
+
                 if self.receive_buffer[channel].empty():
                     del self.receive_buffer[channel]
```
Comment (jberci) [inline channels_redis/core.py]: The lock should also be set to None, otherwise we'd be leaking futures and loops.

Comment (kostko) [reply channels_redis/core.py]: But not here, you mean below when receive count goes to zero?

Comment (jberci) [reply channels_redis/core.py]: Umm... yes, sorry, the locking assert below. I was drilling into this line too, but it was fine, and then I botched the complicated mouse click action.

Comment (kostko) [reply channels_redis/core.py]: Fixed.

**Pull request merged**